precompiles/identity/src/lib.rs

```diff
@@ -0,0 +1,657 @@
+
+		Ok(())
+	}
+
+	#[precompile::public("renameSub(address,(bool, bytes))")]
+	fn rename_sub(handle: &mut impl PrecompileHandle, sub: Address, data: Data) -> EvmResult {
+		let sub = Runtime::Lookup::unlookup(Runtime::AddressMapping::into_account_id(sub.0));
+		let data: pallet_identity::Data = data
+			.try_into()
+			.map_err(|e| RevertReason::custom(e).in_field("data"))?;
+		let call = pallet_identity::Call::<Runtime>::rename_sub { sub, data };
+
+		let origin = Runtime::AddressMapping::into_account_id(handle.context().caller);
+		RuntimeHelper::<Runtime>::try_dispatch(handle, Some(origin).into(), call)?;
+
+		Ok(())
+	}
+
+	#[precompile::public("removeSub(address)")]
+	fn remove_sub(handle: &mut impl PrecompileHandle, sub: Address) -> EvmResult {
+		let sub = Runtime::Lookup::unlookup(Runtime::AddressMapping::into_account_id(sub.0));
+		let call = pallet_identity::Call::<Runtime>::remove_sub { sub };
+
+		let origin = Runtime::AddressMapping::into_account_id(handle.context().caller);
+		RuntimeHelper::<Runtime>::try_dispatch(handle, Some(origin).into(), call)?;
+
+		Ok(())
+	}
+
+	#[precompile::public("quitSub()")]
+	fn quit_sub(handle: &mut impl PrecompileHandle) -> EvmResult {
+		let call = pallet_identity::Call::<Runtime>::quit_sub {};
+
+		let origin = Runtime::AddressMapping::into_account_id(handle.context().caller);
+		RuntimeHelper::<Runtime>::try_dispatch(handle, Some(origin).into(), call)?;
+
+		Ok(())
+	}
+
+	#[precompile::public("identity(address)")]
+	#[precompile::view]
+	fn identity(
+		handle: &mut impl PrecompileHandle,
+		who: Address,
+	) -> EvmResult<Registration<Runtime::MaxAdditionalFields>> {
+		// Storage item: IdentityOf:
+		// BoundedVec((RegistrarIndex(4) + Judgement(17)) * MaxRegistrars(20)) + Balance(16) + IdentityInfo(BoundedVec(Data(33) * MaxAdditionalFields(100)) + Data(33)*7 + 20)
+		handle.record_db_read::<Runtime>(3987)?;
+
+		let who: H160 = who.into();
+		let who = Runtime::AddressMapping::into_account_id(who);
+		let identity = pallet_identity::Pallet::<Runtime>::identity(who);
+
+		Ok(Self::identity_to_output(identity)?)
+	}
+
+	#[precompile::public("superOf(address)")]
+	#[precompile::view]
+	fn super_of(handle: &mut impl PrecompileHandle, who: Address) -> EvmResult<SuperOf> {
+		// Storage item: SuperOf:
+		// AccountId(20) + Data(33)
+		handle.record_db_read::<Runtime>(53)?;
+
+		let who: H160 = who.into();
+		let who = Runtime::AddressMapping::into_account_id(who);
+		if let Some((account, data)) = pallet_identity::Pallet::<Runtime>::super_of(who) {
+			Ok(SuperOf {
+				is_valid: true,
+				account: Address(account.into()),
+				data: Self::convert_data_to_output(data),
+			})
+		} else {
+			Ok(SuperOf::default())
+		}
+	}
+
+	#[precompile::public("subsOf(address)")]
+	#[precompile::view]
+	fn subs_of(handle: &mut impl PrecompileHandle, who: Address) -> EvmResult<SubsOf> {
+		// Storage item: SubsOf:
+		// BoundedVec(AccountId(20) * MaxSubAccounts(100))
+		handle.record_db_read::<Runtime>(200)?;
```

Comment:
Should this be `2000` (`20 * 100`)?

Also, I wonder if there's a way to not overcharge here by something like:

1. Calculate max gas based on upper bound
1. Ensure this much gas is left (without charging)
1. Calculate actual number of accounts
1. Charge gas based on this amount

Reply:
That is a good idea. Are we doing something like that already? I'm unaware if this is possible or maybe was thought through already and agreed upon previously /cc @nanocryk

Reply:
I can open a ticket to revist this, I wasn't able to find something similar that was previously done, and in general I'm unsure that this will be reliable when we've merely checked for the appropriate cost first, which is then followed by multiple actions that do not have this notion of "extra gas I must account for".